### PR TITLE
Add partial stubs for fractions

### PR DESCRIPTION
**Pull request by Michael0x2a**
This commit adds some incomplete stubs for the fractions module.

In particular, this pull request is basically a less complete version of https://github.com/python/typeshed/pull/94 -- this pull request punts on adding type signatures to functions like `__add__` and just sticks with annotating the more obvious functions and methods. This isn't ideal, but does allow us to move forward with creating stubs for the statistics module.

This pull request requires https://github.com/python/typeshed/pull/545 to first be merged and accepted. (For some reason, the stubs for `decimal` in Python 3 are relatively complete, but the stubs for Python 2 are not, which is causing the current test failure).

However, this pull request can be reviewed and accepted independently from my other recent pull request that [refines numbers.pyi](https://github.com/python/typeshed/pull/543). Once that other pull request is accepted, I'm planning on following up with another pull request that adds the `__hash__` function to the `Fractions` class.

### CODE REVIEW

**Comment by gvanrossum** [comment]
Can you deal with the test failure? Seems your stubs are incompatible with
Decimal.

**Comment by Michael0x2a** [comment]
Yeah -- the fix for that is here: https://github.com/python/typeshed/pull/545

(Python 2's stubs for Decimals seem to be much less complete then Python 3's stubs, which is what's causing the incompatibility)

**Comment by gvanrossum** [comment]
I see. But the Python 2 stubs have to be compatible with the crummy
production code we have at Dropbox using it. I think I have to make some
changes there before that can land (or perhaps after).

**Comment by gvanrossum** [comment]
Now that the improved decimal stubs have landed can you rebase and re-push? to trigger the tests?

**Comment by Michael0x2a** [comment]
@gvanrossum -- done! It's green now :)

File: stdlib/2and3/fractions.pyi [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+from numbers import Complex, Rational
+from decimal import Decimal
+import sys
+
+_TComplex = TypeVar('_TComplex', bound=Complex)
+_TOtherNum = TypeVar('_TOtherNum', int, float, Decimal)
+_TNum = TypeVar('_TNum', int, float, Decimal, Complex)
+
+@overload
+def gcd(a: _TComplex, b: _TComplex) -> _TComplex: ...
```
Comment (gvanrossum) [inline stdlib/2and3/fractions.pyi]: I don't know what the docs say, but the module doesn't seem to support complex arguments here.

File: stdlib/2and3/fractions.pyi [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    @classmethod
+    def from_decimal(cls, dec: Decimal) -> 'Fraction': ...
+    def limit_denominator(self, max_denominator: int = 1000000) -> 'Fraction': ...
+
+    @property
+    def numerator(self) -> int: ...
+    @property
+    def denominator(self) -> int: ...
+
+    def __add__(self, other): ...
```
Comment (gvanrossum) [inline stdlib/2and3/fractions.pyi]: Add a # TODO comment here asking for better annotations in the future?

File: stdlib/2and3/fractions.pyi [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    def __pos__(self) -> 'Fraction': ...
+    def __neg__(self) -> 'Fraction': ...
+    def __abs__(self) -> 'Fraction': ...
+    def __trunc__(self) -> int: ...
+    if sys.version_info >= (3, 0):
+        def __floor__(self) -> int: ...
+        def __ceil__(self) -> int: ...
+        def __round__(self, ndigits=None): ...
+
+    # TODO: Once https://github.com/python/typeshed/pull/543 is
```
Comment (gvanrossum) [inline stdlib/2and3/fractions.pyi]: It's been accepted.

File: stdlib/2and3/fractions.pyi (hunk 1 above)
Comment (gvanrossum) [inline stdlib/2and3/fractions.pyi]: At least not in Python 3. It does in Python 2?!

Comment (Michael0x2a) [inline stdlib/2and3/fractions.pyi]: Hmm, so it looks like the return type of gcd here ultimately depends on what the type of `a % b` is, which makes the gcd function more complicated then I was originally expecting.

Some things I observed:

- `numbers.Complex` does not have any kind of dunder methods related to mods, so shouldn't be a legal input type in either Python 2 or Python 3.
- However, the builtins `complex` type _does_ allow you to take the mod in Python 2, but NOT in Python 3.
- If one argument is a complex number, and the other is any other type, the gcd function will either throw a TypeError or will enter an infinite loop (!!).

I also found the full matrix of allowable parameters to return types is this (not taking into account whatever any custom number subclasses might do):

```
            | int          float        Fraction     Decimal      complex
------------+ -----------  -----------  -----------  -----------  -----------
int         | int          float        Fraction     Decimal      Inf. Loop
float       | float        float        float        TypeError    Inf. Loop
Fraction    | Fraction     float        Fraction     TypeError    TypeError
Decimal     | Decimal      TypeError    TypeError    Decimal      TypeError
complex     | Inf. Loop    Inf. Loop    TypeError    TypeError    complex
```

Thankfully, the diagonal (where both parameters are the same type) appear to be relatively sane.

So, given all of the above, and given that `fractions.gcd` is a deprecated function anyways as of Python 3 (people are supposed to use math.gcd instead, which accepts only ints), would you rather I:

- Attempt to come up with a system of overloads to handle the full matrix.
- Attempt to handle the case where both parameters are the same type, and just provide separate annotations for Python 2 and 3 to handle complex numbers.
- Just disallow allowing complex numbers as parameters completely on the grounds that nobody is actually going to do that/they're going to be using cmath instead.
- Don't bother providing a type signature, and just make the input and output types Any.
- Something else?

**Comment by gvanrossum** [comment]
Wow, you're reasoning like pytype. :-)

In 3.5, fractions.gcd() is deprecated in favor of math.gcd(), which only supports integers. I propose to make the type signature do that in all versions. Even the 2.7 docs say "integers"...

**Comment by Michael0x2a** [comment]
Hmm, ok -- I can do that.

I guess the only thing I'm hesitant about is that the source code of `fractions.gcd` explicitly handles the case where neither parameters are ints, which actually makes it bit more general-purpose then `math.gcd`.

But it seems like we're not really going to care about that?

**Comment by Michael0x2a** [comment]
Or actually, maybe make it just accept either int or `numbers.Integral`?

**Comment by gvanrossum** [comment]
Looking at the C source for math.gcd() it seems that supports things with
an __index__ method, which is a proxy for being a sort of integer. So I
think Union[int, Integral] is fine!

**Pull request merged**